Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/TanHelper.kt

```diff
@@ -0,0 +1,33 @@
+package de.rki.coronawarnapp.util
+
+import de.rki.coronawarnapp.ui.submission.TanConstants.MAX_LENGTH
+import java.security.MessageDigest
+import java.util.Locale
+
+object TanHelper {
+    private const val VALID_CHARACTERS = "23456789ABCDEFGHJKMNPQRSTUVWXYZ"
+
+    fun isChecksumValid(tan: String): Boolean {
+        if (tan.trim().length != MAX_LENGTH)
+            return false
+        val subTan = tan.substring(0, MAX_LENGTH - 1).toUpperCase(Locale.getDefault())
+        val tanDigest = MessageDigest.getInstance("SHA-256").digest(subTan.toByteArray())
+        var checkChar = "%02x".format(tanDigest[0])[0]
+        if (checkChar == '0') checkChar = 'G'
+        if (checkChar == '1') checkChar = 'H'
+
+        return checkChar.toUpperCase() == tan.last().toUpperCase()
```

Comment:
here it is locale independent, as it uses `char` type (not easy to see as `var` is used instead of explicit type).